README.md
[![npm version](https://badge.fury.io/js/roboapi.ts.svg)](https://badge.fury.io/js/roboapi.ts)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](LICENSE.md)
# RoboApi.ts

Позволяет использовать интерфейс (платежный шлюз) для оплаты через Робокассу. Поддерживает интеграцию чеков (закон 54-ФЗ)

CHANGELOG.md
# 2.0.0 (2019-11-14)

### Features

* добавил метод getLink ([e3ab249](https://github.com/Altairko/roboapi.ts/commit/e3ab2493bd39e54733fd94370844c5557937138b))


### BREAKING CHANGES

* добавил обязательный параметр с методом хеширования в checkPayment ([e3ab249](https://github.com/Altairko/roboapi.ts/commit/e3ab2493bd39e54733fd94370844c5557937138b))



## 1.0.1 (2019-11-13)


### Bug Fixes

* jsDoc ([d0f2532](https://github.com/Altairko/roboapi.ts/commit/d0f253230b171e3c7faa57a86b6c949c8dbc1342))
* исправил парсинг xml ([2c3bc19](https://github.com/Altairko/roboapi.ts/commit/2c3bc1968396fd8f42a01e8374213382fa1f442a))


### Features

* RobokassaPayAPI класс ([5aa48c7](https://github.com/Altairko/roboapi.ts/commit/5aa48c7ac8c81009bd4bd88d62c4d98d9ec37e9b))
* добавил лицензию и readme ([4faa517](https://github.com/Altairko/roboapi.ts/commit/4faa5172e155baa08c355a79c3efe963ff4cdfc2))
* добавил поле isTest ([d8d9e4d](https://github.com/Altairko/roboapi.ts/commit/d8d9e4dacc5e880d0028f6831919e55c4b5367aa))
* добавил тесты ([a001cd0](https://github.com/Altairko/roboapi.ts/commit/a001cd06b1f860d6807f117cc9020d4ba5dba575))





test/RobokassaPayAPI.spec.ts
import 'mocha';
import {expect} from 'chai';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import RobokassaPayAPI from "../src/RobokassaPayAPI";
import receipt from './helpers/receipt.json'
import {
    OPERATION_COMPLETED_SUCCESSFULLY,
    PARTNER_NOT_AVAILABLE,
    REQUEST_SUCCESSFULLY,
    OPERATION_NOT_FOUND,
    INCORRECT_SIGNATURE,
    FOUND_TWO_OPERATIONS,
    OPERATION_INITIATED,
    OPERATION_CANCELLED_MONEY_NOT_RECEIVED,
    MONEY_CREDITED_STORE,
    MONEY_RETURNED_BUYER,
    OPERATION_SUSPENDED,
    UNKNOWN_ANSWER
} from "../src/helpers/response";
import calcOutSumError from './helpers/calcOutSumError';
import calcOutSumSuccess from './helpers/calcOutSumSuccess';
import fullCurrencies from "./helpers/fullCurrencies";
import emptyCurrencies from "./helpers/emptyCurrencies";
import emptyOperation from "./helpers/emptyOperation";
import fullOperation from "./helpers/fullOperation";
import incompleteOperation from "./helpers/incompleteOperation";

const mrhLogin = 'demo';
const mrhPass1 = 'password_1';
const mrhPass2 = 'password_2';
const isTest = true;


const mock = new MockAdapter(axios);
const sum = 100;
const invId = 93;


describe('RobokassaPayAPI check methods', () => {
    const RobokassaPay = new RobokassaPayAPI({
        mrhLogin,
        mrhPass1,
        mrhPass2,
        isTest
    });

    afterEach(() => {
        mock.reset();
    });

    describe('Получаем ссылку на оплату', () => {
        it('Ссылка успешно получена', () => {
            const paymentLink = RobokassaPay.getLink({
                invId,
                description: 'Тестовая оплата',
                outSum: 100500,
                hashMethod: 'md5'
            });
            expect(paymentLink).to.be.a('string');
        })
    });
    describe('Получаем контрольную сумму подписи', () => {
        it('Получаем строку для подписи без JSON', (done) => {
            const signatureString = RobokassaPay.getSignatureString(sum, invId);
            expect(signatureString).to.equal(`${mrhLogin}:${sum}:${invId}:${mrhPass1}`);
            done();
        });
        it('Получаем строку для подписи с JSON', (done) => {
            const receiptJson = JSON.stringify(receipt);
            const signatureString = RobokassaPay.getSignatureString(sum, invId, receiptJson);
            expect(signatureString).to.equal(`${mrhLogin}:${sum}:${invId}:${receiptJson}:${mrhPass1}`);
            done();
        });
        it('Получаем контрольную сумму без JSON', (done) => {
            const signatureString = RobokassaPay.getSignatureString(sum, invId);
            const signature = RobokassaPay.getSignature(signatureString, 'md5');
            expect(signature).to.equal(
                RobokassaPay.getSignature(
                    `${mrhLogin}:${sum}:${invId}:${mrhPass1}`,
                    'md5'
                )
            );
            done();
        });
        it('Получаем контрольную сумму с JSON', (done) => {
            const receiptJson = JSON.stringify(receipt);
            const signatureString = RobokassaPay.getSignatureString(sum, invId, receiptJson);
            const signature = RobokassaPay.getSignature(signatureString, 'md5');
            expect(signature).to.equal(
                RobokassaPay.getSignature(
                    `${mrhLogin}:${sum}:${invId}:${receiptJson}:${mrhPass1}`,
                    'md5'
                )
            );
            done();
        });
        it('Получаем ошибку при неверном методе хеширования', (done) => {
            const receiptJson = JSON.stringify(receipt);
            const signatureString = RobokassaPay.getSignatureString(sum, invId, receiptJson);
            try {
                // @ts-ignore
                RobokassaPay.getSignature(signatureString, 'md6');
                done('the method of hashing the wrong');
            } catch (e) {
                expect(e).to.throw;
                done();
            }
        })
    });
    describe('Отправка СМС через Robokassa', () => {
        const phone = '71234567890';
        const message = 'Test';
        const smsMock = mock.onGet('https://services.robokassa.ru/SMS/');

        it('СМС успешно отправлено', (done) => {
            smsMock.reply(200, {
                result: true,
                count: 10,
                errorCode: 0,
                errorMessage: ''
            });
            RobokassaPay.sendSms(phone, message).then(response => {
                expect(response, 'В ответе вернулись правильные ключи').to.contain.keys(['code', 'message', 'success']);
                expect(response, 'Вернуло успешно выполненную операцию').to.deep.equal(REQUEST_SUCCESSFULLY);
                done();
            }).catch(e => {
                done(e);
            })
        });
        it('При отправке СМС возникает ошибка', (done) => {
            smsMock.reply(200, {
                result: false,
                count: 10,
                errorCode: 4,
                errorMessage: ''
            });
            RobokassaPay.sendSms(phone, message).then(response => {
                expect(response, 'Вернуло ошибку SMS sending for this partner is not available').to.deep.equal(PARTNER_NOT_AVAILABLE);
                done();
            }).catch(e => {
                done(e);
            })
        });
        it('Некорректный запрос отправки СМС', (done) => {
            smsMock.networkErrorOnce();
            RobokassaPay.sendSms(phone, message).then(() => {
                done('Data should not be obtained');
            }).catch(e => {
                expect(e, 'Правильно обработало исключение').to.throw;
                done();
            })
        });
    });
    describe('Получение суммы с комиссией', () => {
        const commissionMock = mock.onPost('https://auth.robokassa.ru/Merchant/WebService/Service.asmx/CalcOutSumm');
        it('Сумма с комиссией успешно рассчитана', (done) => {
            commissionMock.reply(200, calcOutSumSuccess);
            RobokassaPay.getCommissionSum('', 120).then(result => {
                expect(result).to.equal(120);
                done();
            }).catch(e => {
                done(e);
            });
        });
        it('Некорректная сумма с комиссией', (done) => {
            commissionMock.reply(200, calcOutSumError);
            RobokassaPay.getCommissionSum('', 120).then(() => {
                done(new Error('Комиссия расчитана неверно'));
            }).catch(e => {
                expect(e).to.throw;
                done();
            });
        });
        it('При получении суммы комиссии возникли ошибки', (done) => {
            commissionMock.networkErrorOnce();
            RobokassaPay.getCommissionSum('', 120).then(() => {
                done('Data should not be obtained');
            }).catch(e => {
                expect(e, 'Правильно обработало исключение').to.throw;
                done();
            })
        });
    });
    describe('Получение способов оплаты', () => {
        const currenciesMock = mock.onPost('https://auth.robokassa.ru/Merchant/WebService/Service.asmx/GetCurrencies');
        it('Способы оплаты успешно получены и непустые', (done) => {
            currenciesMock.reply(200, fullCurrencies);
            RobokassaPay.getCurrencies().then(result => {
                result.every(i => expect(i).to.have.all.keys(['Name', 'Label', 'Alias', 'MinValue', 'MaxValue']));
                done();
            }).catch(e => {
                done(e);
            });
        });
        it('Способы оплаты успешно получены и пустые', (done) => {
            currenciesMock.reply(200, emptyCurrencies);
            RobokassaPay.getCurrencies().then(result => {
                expect(result).to.deep.equal([]);
                done();
            }).catch(e => {
                done(e);
            });
        });
        it('При получении способов оплаты возникли ошибки', (done) => {
            currenciesMock.networkErrorOnce();
            RobokassaPay.getCurrencies().then(() => {
                done('Data should not be obtained');
            }).catch(e => {
                expect(e, 'Правильно обработало исключение').to.throw;
                done();
            })
        });
    });
    describe('Получение статуса платежа', () => {
        const checkPaymentMock = mock.onPost('https://auth.robokassa.ru/Merchant/WebService/Service.asmx/OpStateExt');
        it('Статус оплаты получен. Платеж успешно проведен', (done) => {
            checkPaymentMock.reply(200, fullOperation);
            RobokassaPay.checkPayment(1234, 'md5').then(result => {
                expect(result).to.deep.equal(OPERATION_COMPLETED_SUCCESSFULLY);
                done();
            }).catch(e => {
                done(e);
            });
        });
        it('Неверная цифровая подпись запроса', (done) => {
            checkPaymentMock.reply(200, emptyOperation(1));
            RobokassaPay.checkPayment(1234, 'md5').then(result => {
                expect(result).to.deep.equal(INCORRECT_SIGNATURE);
                done();
            }).catch(e => {
                done(e);
            });
        });
        it('Не удалось найти операцию', (done) => {
            checkPaymentMock.reply(200, emptyOperation(3));
            RobokassaPay.checkPayment(1234, 'md5').then(result => {
                expect(result).to.deep.equal(OPERATION_NOT_FOUND);
                done();
            }).catch(e => {
                done(e);
            });
        });
        it('найдено две операции с таким InvoiceID', (done) => {
            checkPaymentMock.reply(200, emptyOperation(4));
            RobokassaPay.checkPayment(1234, 'md5').then(result => {
                expect(result).to.deep.equal(FOUND_TWO_OPERATIONS);
                done();
            }).catch(e => {
                done(e);
            });
        });
        it('Операция только инициализирована', (done) => {
            checkPaymentMock.reply(200, incompleteOperation(5));
            RobokassaPay.checkPayment(1234, 'md5').then(result => {
                expect(result).to.deep.equal(OPERATION_INITIATED);
                done();
            }).catch(e => {
                done(e);
            });
        });
        it('Операция отменена, деньги от покупателя не были получены', (done) => {
            checkPaymentMock.reply(200, incompleteOperation(10));
            RobokassaPay.checkPayment(1234, 'md5').then(result => {
                expect(result).to.deep.equal(OPERATION_CANCELLED_MONEY_NOT_RECEIVED);
                done();
            }).catch(e => {
                done(e);
            });
        });
        it('Деньги от покупателя получены, производится зачисление денег на счет магазина', (done) => {
            checkPaymentMock.reply(200, incompleteOperation(50));
            RobokassaPay.checkPayment(1234, 'md5').then(result => {
                expect(result).to.deep.equal(MONEY_CREDITED_STORE);
                done();
            }).catch(e => {
                done(e);
            });
        });
        it('Деньги после получения были возвращены покупателю', (done) => {
            checkPaymentMock.reply(200, incompleteOperation(60));
            RobokassaPay.checkPayment(1234, 'md5').then(result => {
                expect(result).to.deep.equal(MONEY_RETURNED_BUYER);
                done();
            }).catch(e => {
                done(e);
            });
        });
        it('Исполнение операции приостановлено', (done) => {
            checkPaymentMock.reply(200, incompleteOperation(80));
            RobokassaPay.checkPayment(1234, 'md5').then(result => {
                expect(result).to.deep.equal(OPERATION_SUSPENDED);
                done();
            }).catch(e => {
                done(e);
            });
        });
        it('Неизвестный код сотояния оплаты', (done) => {
            checkPaymentMock.reply(200, incompleteOperation(800));
            RobokassaPay.checkPayment(1234, 'md5').then(result => {
                expect(result).to.deep.equal(UNKNOWN_ANSWER);
                done();
            }).catch(e => {
                done(e);
            });
        });
        it('Неизвестный код результата оплаты', (done) => {
            checkPaymentMock.reply(200, emptyOperation(800));
            RobokassaPay.checkPayment(1234, 'md5').then(result => {
                expect(result).to.deep.equal(UNKNOWN_ANSWER);
                done();
            }).catch(e => {
                done(e);
            });
        });
        it('При получении статуса оплаты возникли ошибки', (done) => {
            checkPaymentMock.networkErrorOnce();
            RobokassaPay.getCurrencies().then(() => {
                done('Data should not be obtained');
            }).catch(e => {
                expect(e, 'Правильно обработало исключение').to.throw;
                done();
            })
        });
    });
});

test/helpers/fullCurrencies.ts
export default `
<?xml version="1.0" encoding="utf-8"?>
<CurrenciesList xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                xmlns="http://merchant.roboxchange.com/WebService/">
    <Result>
        <Code>0</Code>
    </Result>
    <Groups>
        <Group Code="EMoney" Description="Электронным кошельком">
            <Items>
                <Currency Label="Qiwi50RIBRM" Alias="QiwiWallet" Name="QIWI Кошелек"/>
                <Currency Label="YandexMerchantRIBR" Alias="YandexMoney" Name="Яндекс.Деньги"/>
                <Currency Label="WMR30RM" Alias="WMR" Name="WMR"/>
                <Currency Label="W1RIBR" Alias="W1" Name="RUR Единый кошелек" MaxValue="14999"/>
                <Currency Label="ElecsnetWalletRIBR" Alias="ElecsnetWallet" Name="Кошелек Элекснет" MaxValue="14999"/>
            </Items>
        </Group>
        <Group Code="Bank" Description="Через интернет-банк">
            <Items>
                <Currency Label="AlfaBankRIBR" Alias="AlfaBank" Name="Альфа-Клик"/>
                <Currency Label="VTB24RIBR" Alias="VTB24" Name="ВТБ" MinValue="1000"/>
                <Currency Label="W1RIBPSBR" Alias="W1" Name="RUR Единый кошелек" MinValue="1" MaxValue="14999"/>
                <Currency Label="MINBankRIBR" Alias="BankMIN" Name="Московский Индустриальный Банк"/>
                <Currency Label="BSSIntezaRIBR" Alias="BankInteza" Name="Банк Интеза" MaxValue="15000"/>
                <Currency Label="BSSAvtovazbankR" Alias="BankAVB" Name="Банк АВБ"/>
                <Currency Label="FacturaBinBank" Alias="BankBin" Name="БИНБАНК"/>
                <Currency Label="BSSFederalBankForInnovationAndDevelopmentR" Alias="BankFBID"
                          Name="ФБ Инноваций и Развития"/>
                <Currency Label="FacturaSovCom" Alias="BankSovCom" Name="Совкомбанк"/>
                <Currency Label="BSSNationalBankTRUSTR" Alias="BankTrust" Name="Национальный банк ТРАСТ"/>
            </Items>
        </Group>
        <Group Code="BankCard" Description="Банковской картой">
            <Items>
                <Currency Label="BANKOCEAN3R" Alias="BankCard" Name="Банковская карта"/>
                <Currency Label="CardHalvaRIBR" Alias="BankCardHalva" Name="Карта Халва"/>
                <Currency Label="ApplePayRIBR" Alias="ApplePay" Name="Apple Pay"/>
                <Currency Label="SamsungPayRIBR" Alias="SamsungPay" Name="Samsung Pay"/>
            </Items>
        </Group>
        <Group Code="Terminals" Description="В терминале">
            <Items>
                <Currency Label="Qiwi50RIBRM" Alias="QiwiWallet" Name="QIWI Кошелек"/>
                <Currency Label="TerminalsElecsnetRIBR" Alias="TerminalsElecsnet" Name="Терминал Элекснет"
                          MaxValue="15000"/>
            </Items>
        </Group>
        <Group Code="Mobile" Description="Сотовые операторы">
            <Items>
                <Currency Label="MixplatMTSRIBR" Alias="PhoneMTS" Name="МТС" MinValue="10" MaxValue="15000"/>
                <Currency Label="MixplatBeelineRIBR" Alias="PhoneBeeline" Name="Билайн" MinValue="10" MaxValue="15000"/>
                <Currency Label="MixplatTele2RIBR" Alias="PhoneTele2" Name="Tele2" MinValue="10" MaxValue="15000"/>
                <Currency Label="MixplatMegafonRIBR" Alias="PhoneMegafon" Name="Мегафон" MinValue="10"
                          MaxValue="15000"/>
                <Currency Label="CypixTatTelecomRIB" Alias="PhoneTatTelecom" Name="Таттелеком" MinValue="10"
                          MaxValue="15000"/>
            </Items>
        </Group>
        <Group Code="Other" Description="Другие способы">
            <Items>
                <Currency Label="RapidaRIBEurosetR" Alias="StoreEuroset" Name="Евросеть" MaxValue="15000"/>
                <Currency Label="RapidaRIBSvyaznoyR" Alias="StoreSvyaznoy" Name="Связной" MaxValue="15000"/>
                <Currency Label="Biocoin" Alias="BioCoin" Name="BioCoin" MaxValue="15000"/>
            </Items>
        </Group>
    </Groups>
</CurrenciesList>
`;

src/helpers/response.ts
export const REQUEST_SUCCESSFULLY = {
    code: 0,
    message: 'request handled successfully',
    success: true
};
export const NO_QUERY_OPTION = {code: 1, message: 'no query option', success: false};
export const PARTNER_NOT_FOUND = {code: 2, message: 'partner was not found', success: false};
export const PARTNER_INACTIVE = {code: 3, message: 'partner is inactive', success: false};
export const PARTNER_NOT_AVAILABLE = {
    code: 4,
    message: 'SMS sending for this partner is not available',
    success: false
};
export const PARTNER_NOT_POSSIBLE = {
    code: 5,
    message: 'currently sending SMS messages by the indicated partner is not possible',
    success: true
};
export const SMS_LIMIT_HIGHER = {code: 6, message: 'SMS limit is higher', success: false};
export const INVALID_REQUEST_SIGNATURE = {code: 1000, message: 'invalid request signature', success: false};
export const INTERNAL_ERROR = {code: 9999, message: 'internal error', success: false};
export const UNKNOWN_ANSWER = {code: -1, message: 'unknown answer', success: false};
export const OPERATION_INITIATED = {code: 5, message: 'operation initiated', success: false};
export const OPERATION_CANCELLED_MONEY_NOT_RECEIVED = {
    code: 10,
    message: 'operation cancelled money from the buyer was not received',
    success: false
};
export const MONEY_CREDITED_STORE = {
    code: 50,
    message: 'money from the buyer received is made by depositing the money into the account of the store',
    success: false
};
export const MONEY_RETURNED_BUYER = {
    code: 60,
    message: 'money after receipt was returned to the buyer',
    success: false
};
export const OPERATION_SUSPENDED = {code: 80, message: 'operation suspended', success: false};
export const OPERATION_COMPLETED_SUCCESSFULLY = {
    code: 100,
    message: 'operation was completed successfully',
    success: true
};
export const INCORRECT_SIGNATURE = {
    code: 1,
    message: 'Incorrect digital signature request',
    success: false
};
export const OPERATION_NOT_FOUND = {
    code: 3,
    message: 'Couldn\'t find the operation',
    success: false
};
export const FOUND_TWO_OPERATIONS = {
    code: 4,
    message: 'Found two operations with the so InvoiceID',
    success: false
};

src/RobokassaPayAPI.ts
import {join, includes, toUpper, forEach, compact, map} from 'lodash';
import {createHash} from 'crypto';
import axios from 'axios';
import {validate, parse} from 'fast-xml-parser';
import {
    REQUEST_SUCCESSFULLY,
    NO_QUERY_OPTION,
    PARTNER_NOT_FOUND,
    PARTNER_INACTIVE,
    PARTNER_NOT_AVAILABLE,
    PARTNER_NOT_POSSIBLE,
    SMS_LIMIT_HIGHER,
    INVALID_REQUEST_SIGNATURE,
    INTERNAL_ERROR,
    UNKNOWN_ANSWER,
    OPERATION_INITIATED,
    OPERATION_CANCELLED_MONEY_NOT_RECEIVED,
    MONEY_CREDITED_STORE,
    MONEY_RETURNED_BUYER,
    OPERATION_SUSPENDED,
    OPERATION_COMPLETED_SUCCESSFULLY,
    INCORRECT_SIGNATURE, OPERATION_NOT_FOUND, FOUND_TWO_OPERATIONS
} from "./helpers/response";

export type TSimpleObj = {
    [key: string]: any
}
export type TCurrency = 'USD' | 'EUR' | 'KZT' | null;
export type TPay = {
    Name: string;
    Label: string;
    Alias: string;
    MinValue: number;
    MaxValue: number;
}
export type THash = 'md5' | 'ripemd160' | 'sha1' | 'sha256' | 'sha384' | 'sha512' | 'md5';
export type TResponse = {
    code: number;
    message: string;
    success: boolean;
}
export type TConstructorArgs = {
    mrhLogin: string,
    mrhPass1: string,
    mrhPass2: string,
    outCurrency?: TCurrency,
    isTest?: boolean
}
export type THttpMethod = 'GET' | 'POST';
export type TLinkArgs = {
    /** Сумма платежа */
    outSum: number;

    /** Номер инвойса в магазине */
    invId: number;

    /** Описание */
    description: string;

    /** Метод хеширования */
    hashMethod: THash;

    /** Пользовательские параметры */
    customSettings?: TSimpleObj;
}

class RobokassaPayAPI {

    /**
     * @var string
     */
    private readonly mrhLogin: string;

    /**
     * @var string
     */
    private readonly mrhPass1: string;

    /**
     * @var string
     */
    private readonly mrhPass2: string;

    /**
     * @var string
     */
    private readonly apiUrl: string;

    /**
     * @var string
     */
    private readonly paymentUrl: string;

    /**
     * @var string
     */
    private readonly smsUrl: string;

    /**
     * @var TCurrency
     */
    private readonly outCurrency: TCurrency;

    /**
     * @var boolean
     */
    private readonly isTest: boolean;

    /**
     *
     * @param {THttpMethod} httpMethod
     * @param {string} url
     * @param {TSimpleObj} params
     *
     * @return Promise<any>
     */
    private static createRequest(httpMethod: THttpMethod, url: string, params: TSimpleObj): Promise<any> {
        return axios({
            method: httpMethod,
            url,
            [httpMethod === 'GET' ? 'params' : 'data']: params,
        });
    }

    /**
     * @param {TConstructorArgs} args
     * @param {string} args.mrhLogin - идентификатор
     * @param {string} args.mrhPass1 - пароль 1
     * @param {string} args.mrhPass2 - пароль 2
     * @param {TCurrency} args.outCurrency - валюта (если не RUB)
     * @param {boolean} args.isTest - тестовый запрос
     */
    constructor(args: TConstructorArgs) {
        const {mrhLogin, mrhPass1, mrhPass2, outCurrency, isTest} = args;
        this.mrhLogin = mrhLogin;
        this.mrhPass1 = mrhPass1;
        this.mrhPass2 = mrhPass2;
        this.outCurrency = outCurrency || null;
        this.apiUrl = 'https://auth.robokassa.ru/Merchant/WebService/Service.asmx';
        this.smsUrl = 'https://services.robokassa.ru/SMS/';
        this.paymentUrl = 'https://auth.robokassa.ru/Merchant/Index.aspx';
        this.isTest = isTest || false;
    }

    /**
     * Возвращает ссылку для перехода на страницу с оплатой
     * @param {TLinkArgs} args - аргументы
     * @return string
     */
    public getLink(args: TLinkArgs): string {
        const {outSum, invId, description, hashMethod, customSettings} = args;

        let customSettingsStr;
        if (customSettings) {
            customSettingsStr = map(customSettings, (value, key) => {
                return `SHP_${key}=${value}`
            }).join(':')
        }

        const signature = this.getSignature(compact([
            this.mrhLogin,
            outSum,
            invId,
            this.mrhPass1,
            customSettingsStr
        ]).join(':'), hashMethod);

        const requestParams = compact([
            `MerchantLogin=${this.mrhLogin}`,
            `OutSum=${outSum}`,
            `InvoiceID=${invId}`,
            `Description=${description}`,
            `SignatureValue=${signature}`,
            `IsTest=${Number(this.isTest)}`,
            customSettingsStr
        ]).join('&');

        return `${this.paymentUrl}?${requestParams}`
    }

    /**
     * @param {THttpMethod} httpMethod
     * @param {string} apiMethod
     * @param {TSimpleObj}  params
     *
     * @return Promise<TSimpleObj>
     */
    private getData(httpMethod: THttpMethod, apiMethod: string, params: TSimpleObj): Promise<TSimpleObj> | never {
        return this.parseXmlAndConvertToJson(httpMethod, this.apiUrl, apiMethod, params);
    }

    /**
     * Если receiptJson пустой - то в формировании сигнатуры
     * он не будет использоваться, а если не пустой - используем его json-представление
     *
     * @param {number} sum
     * @param {number} invId - Номер счета в магазине
     * @param {string} receiptJson - В этом параметре передается информация о перечне товаров/услуг, количестве,
     * цене и ставке налога по каждой позиции.
     *
     * @return string
     */
    public getSignatureString(sum: number, invId: number, receiptJson: string = ''): string {
        return join(compact([this.mrhLogin, sum, invId, this.outCurrency, receiptJson, this.mrhPass1]), ':');
    }

    /**
     * Генерирует хеш для строки string с помощью метода method
     *
     * @param {string} string
     * @param {string} method
     *
     * @return string | never
     *
     * @throws \Exception
     */
    public getSignature(string: string, method: THash = 'md5'): string | never {
        const methods = ['md5', 'ripemd160', 'sha1', 'sha256', 'sha384', 'sha512'];
        if (includes(methods, method)) {
            return toUpper(createHash(method).update(string).digest('hex'));
        }

        throw new Error('Wrong Signature Method');
    }

    /**
     * Отправляет СМС с помощью GET-запроса на робокассу
     *
     * @param {string} phone
     * @param {string} message
     *
     * @return boolean | never
     * @throws \Exception
     */
    public async sendSms(phone: string, message: string): Promise<TResponse> | never {
        const params = {
            phone,
            message,
            login: this.mrhLogin,
            signature: this.getSignature(`${this.mrhLogin}:${phone}:${message}:${this.mrhPass1}`),
        };

        try {
            const response = await RobokassaPayAPI.createRequest('GET', this.smsUrl, params);
            switch (response?.data?.errorCode) {
                case 0:
                    return REQUEST_SUCCESSFULLY;
                case 1:
                    return NO_QUERY_OPTION;
                case 2:
                    return PARTNER_NOT_FOUND;
                case 3:
                    return PARTNER_INACTIVE;
                case 4:
                    return PARTNER_NOT_AVAILABLE;
                case 5:
                    return PARTNER_NOT_POSSIBLE;
                case 6:
                    return SMS_LIMIT_HIGHER;
                case 1000:
                    return INVALID_REQUEST_SIGNATURE;
                case 9999:
                    return INTERNAL_ERROR;
                default:
                    return UNKNOWN_ANSWER;
            }
        } catch (error) {
            throw error;
        }
    }

    /**
     * Возвращает сумму к оплате с учетом комиссии.
     *
     * @param {string} incCurrLabel Кодовое имя метода оплаты
     * @param {int}    sum          Стоимость товара
     *
     * @return Promise<number> Стоимость, которую необходимо передавать в Робокассу.
     */
    public async getCommissionSum(incCurrLabel: string, sum: number): Promise<number> | never {
        const parsed = await this.getData('POST', 'CalcOutSumm', {
            MerchantLogin: this.mrhLogin,
            IncCurrLabel: incCurrLabel,
            IncSum: sum,
            IsTest: Number(this.isTest)
        });

        try {
            if (parsed?.CalcSummsResponseData?.OutSum > 0) {
                return parsed?.CalcSummsResponseData?.OutSum
            }

            throw new Error('there was an error in the calculation of the commission');
        } catch (e) {
            throw e;
        }
    }

    /**
     * Получение списка способов оплаты (в офф. документации валют), доступных магазину
     *
     * @return Promise<TPay[] | unknown[]> | never
     */
    public async getCurrencies(): Promise<TPay[] | unknown[]> | never {
        const parsed = await this.getData('POST', 'GetCurrencies', {
            MerchantLogin: this.mrhLogin,
            Language: 'ru',
            IsTest: Number(this.isTest)
        });

        try {
            const paymentMethods: TPay[] | unknown[] = [];

            if (parsed?.CurrenciesList?.Groups) {
                forEach(parsed?.CurrenciesList?.Groups, group => {
                    if (group?.[0]?.Items?.Currency) {
                        forEach(group?.[0]?.Items?.Currency, currency => {
                            paymentMethods.push({
                                Name: currency['@_Name'],
                                Label: currency['@_Label'],
                                Alias: currency['@_Alias'],
                                MinValue: currency['@_MinValue'] ? Number(currency['@_MinValue']) : 0,
                                MaxValue: currency['@_MaxValue'] ? Number(currency['@_MaxValue']) : 9999999,
                            } as TPay);
                        });
                    }
                })
            }

            return paymentMethods;
        } catch (e) {
            throw e
        }
    }

    /**
     * Парсит XML в JSON
     *
     * @param {THttpMethod} httpMethod
     * @param {string} apiUrl
     * @param {string} controller
     * @param {TSimpleObj} params
     *
     * @return Promise<TSimpleObj> | never
     */
    private parseXmlAndConvertToJson(
        httpMethod: THttpMethod,
        apiUrl: string,
        controller: string,
        params: TSimpleObj
    ): Promise<TSimpleObj> | never {
        const url = `${apiUrl}/${controller}`;
        return RobokassaPayAPI.createRequest(httpMethod, url, params).then(response => {
            if (response?.data && validate(response?.data)) {
                return parse(response.data, {
                    ignoreAttributes: false
                })
            }
        }).catch(error => {
            return Promise.reject(error.message)
        });
    }

    /**
     * Запрашивает у робокассы подтверждение платежа
     *
     * @param {int} invId
     * @param {THash} hashMethod
     *
     * @return Promise<TResponse> | never
     */
    public async checkPayment(invId: number, hashMethod: THash): Promise<TResponse> | never {
        const result = await this.getData('POST', 'OpStateExt', {
            MerchantLogin: this.mrhLogin,
            InvoiceID: invId,
            Signature: this.getSignature(`${this.mrhLogin}:${invId}:${this.mrhPass2}`, hashMethod),
            IsTest: Number(this.isTest)
        });

        try {
            switch (result?.OperationStateResponse?.Result?.Code) {
                case 0:
                    switch (result?.OperationStateResponse?.State?.Code) {
                        case 5:
                            return OPERATION_INITIATED;
                        case 10:
                            return OPERATION_CANCELLED_MONEY_NOT_RECEIVED;
                        case 50:
                            return MONEY_CREDITED_STORE;
                        case 60:
                            return MONEY_RETURNED_BUYER;
                        case 80:
                            return OPERATION_SUSPENDED;
                        case 100:
                            return OPERATION_COMPLETED_SUCCESSFULLY;
                        default:
                            return UNKNOWN_ANSWER;
                    }
                case 1:
                    return INCORRECT_SIGNATURE;
                case 3:
                    return OPERATION_NOT_FOUND;
                case 4:
                    return FOUND_TWO_OPERATIONS;
                default:
                    return UNKNOWN_ANSWER;
            }
        } catch (e) {
            throw e
        }
    }
}

export default RobokassaPayAPI;